Feature journal aggregation

Azure7111 opened the pull request:
This PR aims to provide means to aggregate journal entries. This should then be used later on to combine e.g. updating the properties of a work package and commenting on it into a single journal. Aggregated journals should never be written.

https://community.openproject.org/work_packages/20686

Commit:
First incomplete draft

Commit:
improved draft

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://azure7111-feature-journal-aggregation.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://azure7111-feature-journal-aggregation.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

Azure7111 commented:
Please note that this PR is open now mainly to make sure I haven't forgotten something. So if you miss anything please don't hesitate to comment. Tests will follow ;)

Azure7111 commented on app/models/journal_aggregator.rb:
```diff
@@ -0,0 +1,92 @@
+    AggregatedJournal.new(journal_a, journal_b)
+  end
+
+  def self.are_mergeable?(journal_a, journal_b)
```
I'd be glad for any suggestion to improve this method.

NobodysNightmare replied:
For general readability I would do two things:
1. extract private methods for all conditions
   - this spares you the comments above the if statements
   - and makes the code easier readable
2. I would like to see whether joining such methods into a single boolean expression could still be a concise and readable way of writing this method

e.g.:

``` ruby
return true if journal_a.equal? journal_b

same_journable?(journal_a, journal_b) &&
  same_author?(journal_a, journal_b) &&
  journals_are_consecutive?(journal_a, journal_b) &&
  ...
```

Overall this should also decrease the ABC score, but I don't know if it is enough.

Commit:
User proper attribute and fix condition

NobodysNightmare commented on app/models/journal_aggregator.rb:
```diff
@@ -0,0 +1,92 @@
+    end
+
+    # The journals need to be consecutive
+    if (journal_a.id - journal_b.id).abs != 1
```
I don't like this for multiple reasons:
- you assume to know the rule by which ids are generated
- It is sufficient if they are consecutive w.r.t. to their work package, isn't it? But the id will also increase when I change a different work package

Azure7111 replied:
`journal.version` might be better then, right?

NobodysNightmare replied:
Yes, after checking `journal.rb` I would agree that this is the right attribute.

Azure7111 replied:
Would you deem the comment to be OK for that case?

NobodysNightmare replied:
As stated above and stated in a talk yesterday:

I think comments of this kind are completely unneccessary, because extracting a method can serve the purpose of explaining what is done, e.g. `are_journals_consecutive?(journal_a, journal_b)`.

The upside here is that method names do not [rot](http://blogs.msdn.com/b/ericlippert/archive/2004/05/04/125893.aspx) as fast as comments do...

NobodysNightmare commented on app/models/journal_aggregator.rb:
```diff
@@ -0,0 +1,92 @@
+    end
+
+    # Never merge comments
+    if !nil_or_empty?(journal_a.notes) && !nil_or_empty?(journal_b.notes)
```
This is one of the valid use cases for `blank?` and `present?`.

NobodysNightmare commented on app/models/journal_aggregator.rb:
```diff
@@ -0,0 +1,92 @@
+      return false
+    end
+
+    # Never merge comments
```
This comment is wrong or at least easy to misunderstand. We merge comments to other journals, we just don't merge if we had to merge comments on both sides.

Azure7111 replied:
So how about "Never merge two non-empty comments" then?

NobodysNightmare replied:
Actually I would leave the comment out entirely... The semantics of "don't merge if both have a comment" are quite obvious from the code.
What you _could_ explain, is why we need to ensure this (because we don't want to merge texts and get problems editing the comment, bla bla bla).
But I deem this explanation optional

NobodysNightmare commented on app/models/journal_aggregator.rb:
```diff
@@ -0,0 +1,92 @@
+      aggregated_journals.combination(2).each do |journal_a, journal_b|
+        if are_mergeable?(journal_a, journal_b)
+          result + [merge(journal_a, journal_b)]
+          result - [journal_a, journal_b]
```
`+=` and `-=`?

Azure7111 replied:
Yep, and the mere fact that you ask is either an indicator for the fact that this code needs refactoring as it is hard to understand or at the very least needs some explanation via comment.

Azure7111 replied:
The idea is to keep the merged journals and remove the journals that got merged, then keep on looking whether the merged journals can be combined any further.

NobodysNightmare replied:
We talked about this in person: My comment was about the mere fact that you never assign the return value of `+` and `-` back to `result` and that you might want to do that.

The purpose of the algorithm was clear enough.

NobodysNightmare commented on app/models/journal_aggregator.rb:
```diff
@@ -0,0 +1,92 @@
+    true
+  end
+
+  private
```
Hound is right: The modifier is useless, because only methods in a different scope are defined afterwards.

Since oyu only plan on doing class methods here, I would suggest you to wrap everything into a `class << self`, e.g.:

``` ruby
class Foo
  def self.bar; end
  def self.baz; end
end
```

becomes

``` ruby
class Foo
  class << self
    def bar; end
    def baz; end
  end
end
```

What that does is it redefines self to be the class and therefore all method definitions inside are automatically applied to the class instead of the instance side. It also allows you to use `private` and get the desired result.

Azure7111 commented:
Thinking about this there's at least one main point still missing here, which is the comparison of attributes from `aggregated_journal` to anything else. Like `created_at`, `id` and probably some more.

The main challenge here will be to find a somewhat generic way.

NobodysNightmare commented on app/models/journal/aggregated_journal.rb:
```diff
@@ -0,0 +1,50 @@
+class AggregatedJournal
+  def initialize(*journals)
+    journals.combination(2).each do |journal_a, journal_b|
+      unless JournalAggregator.are_mergeable?(journal_a, journal_b)
```
I don't like the link back to the `JournalAggregator` here.

Does the `AggregatedJournal` really have to care if the merge was valid?

Azure7111 replied:
Shouldn't an `AggregatedJournal` still contain valid data? o0

Though I have already noticed that the check for `are_mergeable?` might happen a tad too often, but I'm not quite sure if there is a good place to remove it.

I'm also happy to move methods between classes, since I don't think this might be the perfect fit yet.

NobodysNightmare replied:
It is arguable that `AggregatedJournal` is a dumb data holding class, that will work correctly if being used correctly. But I can't cite any book here and opinions may vary.

It should be harmless in that the `JournalAggregator` surely knows how to do its job, the `AggregatedJournals` could then rely on the hope that everyone creating them knows what he does.

But that is only my opinion... defensive coding would probably say that you are doing it the right way, don't know...

Azure7111 replied:
I tend to like the argument you brought forth, because it will also make it clear that this class should not care about mergeability.

NobodysNightmare commented on app/models/journal_aggregator.rb:
```diff
@@ -0,0 +1,92 @@
+    advance = journals_array.count == 0
+    while advance
+      result = aggregated_journals.dup
+      aggregated_journals.combination(2).each do |journal_a, journal_b|
```
For the start of this PR we might want to keep this loop for aggregation to test other parts of the code, however I see some problems:
1. aggregation will not work for interleaved aggregatable journals
2. I'd wish for a single pass algorithm instead of the multi pass approach that you chose, however, I am not yet sure whether this is possible

To explain on the interleaving, lets assume the following journals:

```
journal(wp: 74, author: A, changes: { ... })
journal(wp: 75, author: B, changes: { ... })
journal(wp: 74, author: A, changes: { ... })
```

In this case it is perfectly valid to merge the first and the third journal, because w.r.t. their work package they are consecutive and they are by the same author. However, the `combination(2)` iteration ensures that they will never meet and therefore not be merged.

Azure7111 replied:
I am not sure if I agree with your first point. To my understanding [combination(2)](http://ruby-doc.org/core-2.1.5/Array.html#method-i-combination) takes an array an creates a new one which contains every possible pairing from the elements from the former array. An since the order within the pairs should not make a difference I have to object to your first point.

NobodysNightmare replied:
Oh okay, I just RTFM and have to agree. I expected different behaviour.

Then my only objection is that we need to consider that this method carries a high algorithmic complexity:

```
pry(main)> (1..5).to_a.combination(2).count
=> 10
pry(main)> (1..10).to_a.combination(2).count
=> 45
pry(main)> (1..20).to_a.combination(2).count
=> 190
```

Note: I am not saying the complexity is unjustified, but we probably want to think about it twice.

Commit:
Add some initial journal aggregator specs

Commit:
Fix notes collection

Commit:
Add some initial aggregated journal specs

NobodysNightmare committed:
rewrite merge loop

For now the implementation takes O(n!), but this will improve vastly by existing the loop early

NobodysNightmare committed:
break out of merge loop early if applicable

- only if there is no further aggregation possible for the current WP
- should reduce real world complexity a lot (because of break after temporal distance exceeded)

NobodysNightmare committed:
iterate over a duplicate of the journal queue

I first suspected I could safely (without missing anything) iterate over an array while deleting from it,
which I would have loved ruby for ^^

But as it turns out, it is not: http://stackoverflow.com/questions/31294630

NobodysNightmare committed:
use class << self

- avoiding "self." prefix on every def
- allowing the use of "private"

NobodysNightmare committed:
replace own helper by "present?"

NobodysNightmare commented:
Finally closing this one in favor of https://github.com/opf/openproject/pull/3230
